server/auth/auth.go

```diff
@@ -24,18 +24,270 @@ import (
+		return nil, errs.ErrUserNotFound.FastGenByArgs(name)
+	}
+
+	return user, nil
+}
+
+// GetUsers returns all available roles.
+func (m *userManager) GetUsers() map[string]*User {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	return m.users
+}
+
+// CreateUser creates a new user.
+func (m *userManager) CreateUser(name string, password string) error {
+	_, err := m.GetUser(name)
+	if err == nil {
+		return errs.ErrUserExists.GenWithStackByArgs(name)
+	}
+
+	user, err := NewUser(name, GenerateHash(password))
+	if err != nil {
+		return err
+	}
+
+	userJSON, err := json.Marshal(user)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	userPath := path.Join(userPrefix, name)
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Add user to kv.
+	err = m.kv.Save(userPath, string(userJSON))
+	if err != nil {
+		return err
+	}
+
+	// Add user to memory cache.
+	m.users[name] = user
+
+	return nil
+}
+
+// DeleteUser deletes a user.
+func (m *userManager) DeleteUser(name string) error {
+	_, err := m.GetUser(name)
+	if err != nil {
+		return err
+	}
+
+	userPath := path.Join(userPrefix, name)
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Delete user from kv.
+	err = m.kv.Remove(userPath)
+	if err != nil {
+		return err
+	}
+
+	// Delete user from memory cache.
+	delete(m.users, name)
+
+	return nil
+}
+
+// ChangePassword changes password of a user.
+func (m *userManager) ChangePassword(name string, password string) error {
+	user, err := m.GetUser(name)
+	if err != nil {
+		return err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	user.Hash = GenerateHash(password)
+
+	// Update user in kv
+	userJSON, err := json.Marshal(user)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	userPath := path.Join(userPrefix, name)
+
+	err = m.kv.Save(userPath, string(userJSON))
+	if err != nil {
+		return err
+	}
+
+	// No need to update user in memory cache again.
+	return nil
+}
+
+// SetRoles sets roles of a user.
+func (m *userManager) SetRoles(name string, roles []string) error {
+	user, err := m.GetUser(name)
+	if err != nil {
+		return err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	user.RoleKeys = roles
+
+	// Update user in kv
+	userJSON, err := json.Marshal(user)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	userPath := path.Join(userPrefix, name)
+
+	err = m.kv.Save(userPath, string(userJSON))
+	if err != nil {
+		return err
+	}
+
+	// No need to update user in memory cache again.
+	return nil
+}
+
+// AddRole adds a role to a user.
+func (m *userManager) AddRole(name string, role string) error {
+	user, err := m.GetUser(name)
+	if err != nil {
+		return err
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if ok := user.appendRole(role); !ok {
+		return errs.ErrUserHasRole.FastGenByArgs(name, role)
+	}
+
+	userJSON, err := json.Marshal(role)
+	if err != nil {
+		return errs.ErrJSONMarshal.Wrap(err).GenWithStackByCause()
+	}
+	userPath := path.Join(userPrefix, name)
```

Comment:
I think we can define a function in util like `GetUserPath` and `GetRolePath`.

Reply:
Extracted to util.